File: src/projects.ts
import * as A from "@automerge/automerge";
import { randomUUIDv7 } from "bun";
import type { Project, Workspace } from "./types";

export function addProject(
  doc: A.Doc<Workspace>,
  name: string,
  secretIds: string[] = []
): A.Doc<Workspace> {
  const id = randomUUIDv7();
  return A.change(doc, `add project ${name}`, (d) => {
    d.projects[id] = { id, name, secret_ids: secretIds };
  });
}

export function removeProject(
  doc: A.Doc<Workspace>,
  id: string
): A.Doc<Workspace> {
  return A.change(doc, `remove project ${id}`, (d) => {
    delete d.projects[id];
  });
}

export function addSecretToProject(
  doc: A.Doc<Workspace>,
  projectId: string,
  secretId: string
): A.Doc<Workspace> {
  return A.change(doc, `add secret ${secretId} to project ${projectId}`, (d) => {
    const project = d.projects[projectId];
    if (!project) throw new Error(`Project ${projectId} not found`);
    project.secret_ids.push(secretId);
  });
}

export function updateProject(
  doc: A.Doc<Workspace>,
  id: string,
  name: string
): A.Doc<Workspace> {
  return A.change(doc, `update project ${id}`, (d) => {
    const p = d.projects[id];
    if (!p) throw new Error(`Project ${id} not found`);
    p.name = name;
  });
}

export function setProjectSecrets(
  doc: A.Doc<Workspace>,
  projectId: string,
  secretIds: string[]
): A.Doc<Workspace> {
  return A.change(doc, `set secrets for project ${projectId}`, (d) => {
    const p = d.projects[projectId];
    if (!p) throw new Error(`Project ${projectId} not found`);
    p.secret_ids.splice(0, p.secret_ids.length, ...secretIds);
  });
}

export function listProjects(doc: A.Doc<Workspace>): Project[] {
  return Object.values(doc.projects);
}

File: envi/envilib/src/projects.rs
use crate::{
    error::{Error, Result},
    types::{EnviDocument, Project},
};
use automerge::AutoCommit;
use autosurgeon::{hydrate, reconcile};
use uuid::Uuid;

pub fn add_project(doc: &mut AutoCommit, name: &str) -> Result<()> {
    let id = Uuid::now_v7().to_string();
    let mut state: EnviDocument = hydrate(doc)?;
    state.projects.insert(
        id.clone(),
        Project {
            id,
            name: name.to_string(),
            secret_ids: vec![],
        },
    );
    reconcile(doc, &state)?;
    Ok(())
}

pub fn remove_project(doc: &mut AutoCommit, id: &str) -> Result<()> {
    let mut state: EnviDocument = hydrate(doc)?;
    state.projects.remove(id);
    reconcile(doc, &state)?;
    Ok(())
}

pub fn update_project(doc: &mut AutoCommit, id: &str, name: &str) -> Result<()> {
    let mut state: EnviDocument = hydrate(doc)?;
    let project = state
        .projects
        .get_mut(id)
        .ok_or_else(|| Error::ProjectNotFound(id.to_string()))?;
    project.name = name.to_string();
    reconcile(doc, &state)?;
    Ok(())
}

pub fn set_project_secrets(
    doc: &mut AutoCommit,
    project_id: &str,
    secret_ids: Vec<String>,
) -> Result<()> {
    let mut state: EnviDocument = hydrate(doc)?;
    let project = state
        .projects
        .get_mut(project_id)
        .ok_or_else(|| Error::ProjectNotFound(project_id.to_string()))?;
    project.secret_ids = secret_ids;
    reconcile(doc, &state)?;
    Ok(())
}

pub fn list_projects(doc: &AutoCommit) -> Result<Vec<Project>> {
    let state: EnviDocument = hydrate(doc)?;
    Ok(state.projects.into_values().collect())
}
